crates/sui-node/src/lib.rs

```diff
@@ -797,14 +798,17 @@ impl SuiNode {
             } else {
                 self.reconfigure_state(next_epoch).await;
 
-                if self.state.is_validator() {
+                let new_epoch_store = self.state.load_epoch_store_one_call_per_task();
+                assert_eq!(new_epoch_store.epoch(), next_epoch);
+
+                if self.state.is_validator(&new_epoch_store) {
                     info!("Promoting the node from fullnode to validator, starting grpc server");
 
                     Some(
                         Self::construct_validator_components(
                             &self.config,
                             self.state.clone(),
-                            self.state.epoch_store().clone(),
+                            new_epoch_store.clone(),
```

Comment:
Ok, I had to do a lot of reading to understand how this could possibly be ok.

We only call construct_validator_components in two places: start up, and fullnode->validator promotion (currently unexercised). And the epoch store is only used in ConsensusAdapter::submit_recovered(), and not retained.

So, two things: 
- this is really confusing
- I think its probably not correct to call submit_recovered() in the fullnode promotion case.

I think instead we should remove the call to submit_recovered from ConsensusAdapter::new(), and instead call it explicitly only in the validator startup case in this module.

Reply:
Will address this in a separate PR